content/blog/elastic-apm-plugin.md

```diff
@@ -0,0 +1,113 @@
+
+```java
+@Override
+public ElementMatcher<? super TypeDescription> getTypeMatcher() {
+    return named("io.r2dbc.postgresql.PostgresqlStatement");
+}
+
+@Override
+public ElementMatcher<? super MethodDescription> getMethodMatcher() {
+    return named("execute").and(takesArgument(0, named("java.lang.String")));
+}
+```
+
+We can see we match on 2 things here, the class name of the type we want to match and the method we want to match on. Now we need to define what we need to do when that method is called:
+
+```java
+@Override
+public String getAdviceClassName() {
+    return "nl.skyworkz.apm.agent.r2dbc.postgresql.R2dbcPostgresqlInstrumentation$HandleExecuteStatementAdvice";
+}
+
+public static class HandleExecuteStatementAdvice {
+
+    private static final SignatureParser signatureParser = new SignatureParser();
+
+    @Advice.OnMethodExit(inline = false)
+    @Advice.AssignReturned.ToReturned(typing = Assigner.Typing.DYNAMIC)
+    public static Flux<PostgresqlResult> onExitExecute(@Advice.Argument(0) String sql,
+                                                       @Advice.Return Flux<io.r2dbc.postgresql.api.PostgresqlResult> result) {
+        return Mono.defer(() -> Mono.just(createSpan(sql)))
+                .flatMapMany(spanBuilder -> {
+                    Span span = spanBuilder.startSpan();
+
+                    return result
+                            .doOnComplete(() -> {
+                                span.setStatus(StatusCode.OK).end();
+                            })
+                            .doOnError(throwable -> {
+                                span.recordException(throwable).end();
+                            });
+                });
+    }
+
+    /**
+     * See https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/semantic_conventions/database.md
+     */
+    private static SpanBuilder createSpan(String sql) {
+        StringBuilder signature = new StringBuilder();
+        signatureParser.querySignature(sql, signature, false);
+
+        return getTracer("r2dbc-postgresql")
+                .spanBuilder(signature.toString())
+                .setSpanKind(SpanKind.CLIENT)
+                .setAttribute("db.system", "postgresql")
+                .setAttribute("db.statement", sql);
+    }
+}
+```
+
+We wrap around the Reactor type here to start the span before we subscribe to the query result and once it completes we close the span. The span contains the statement (query) being executed. The query will now show up in our timeline (like in the screenshot earlier) and we'll be able to see a summary of the statement and how long it took to execute that statement.
+
+The signature parser that is being used is a copy of the `SignatureParser` class from the JDBC plugin of the APM agent[^5]. To package the plugin a few more steps are needed, you can find them on the Elastic website: https://www.elastic.co/guide/en/apm/agent/java/current/plugin-api.html
+
+
```

Comment:
Sadly, these links are not playing very nicely. Is there an alternative way of linking to these articles?